Remove memory allocation for ELF sections

LambertWSJ opened the pull request:
Since the section headers of ELF are continuous, all headers can be accessed by getting the starting address, so there is no need to allocate memory for this.

jserv reviewed (changes requested):
Eliminate the warnings reported by Codacy/Cppcheck.

jserv commented:
> Eliminate the warnings reported by Codacy/Cppcheck.

Reported by Cppcheck:
```
tools/rv_histogram.c:192:31: style: Parameter 'ir' can be declared as pointer to const [constParameterPointer]
void reg_hist_incr(rv_insn_t *ir)
                              ^
tools/rv_histogram.c:217:32: style: Parameter 'ir' can be declared as pointer to const. However it seems that 'insn_hist_incr' is a callback function, if 'ir' is declared with const you might also need to cast function pointer(s). [constParameterCallback]
void insn_hist_incr(rv_insn_t *ir)
                               ^
tools/rv_histogram.c:235:46: note: You might need to cast the function pointer here
    hist_record = show_reg ? reg_hist_incr : insn_hist_incr;
                                             ^
tools/rv_histogram.c:217:32: note: Parameter 'ir' can be declared as pointer to const
void insn_hist_incr(rv_insn_t *ir)
                               ^
tools/rv_histogram.c:253:28: style: Variable 'shdr' can be declared as pointer to const [constVariablePointer]
        struct Elf32_Shdr *shdr = (struct Elf32_Shdr *) shdrs + i;
                           ^
tools/rv_histogram.c:260:20: style: Variable 'exec_end_addr' can be declared as pointer to const [constVariablePointer]
        uintptr_t *exec_end_addr =
                   ^
```

@LambertWSJ, you may submit pull requests to eliminate the warnings reported by Cppcheck later.

LambertWSJ committed:
Remove memory allocation for ELF sections

* Since the section headers of ELF are continuous, all headers can be
accessed by getting the starting address, so there is no need to
allocate memory for this.

* Eliminate MISRA C warnings as possible and follow contribution guide
* Remove `getopt` and refactor `parse_args`, but the code is not elegant
enough

jserv commented on tools/rv_histogram.c:
```diff
@@ -189,39 +194,38 @@ void print_hist_stats(const rv_hist_t *stats, size_t stats_size)
 {
-    if (!ir)
-        return;
+    if (ir != NULL) {
```
Don't change the function body.

jserv replied:
Check https://github.com/sysprog21/rv32emu/blob/master/CONTRIBUTING.md carefully: Avoid unnecessary nesting levels.

LambertWSJ replied:
Modifications similar to the single-line if statement have been restored.

jserv commented on tools/rv_histogram.c:
```diff
@@ -241,39 +245,35 @@
 #endif
             insn = *((uint32_t *) ptr);
-            ptr = (uintptr_t *) ((uint8_t *) ptr + 4);
+            ptr = ptr + 4;
```
Rewrite as `ptr += 4`.

jserv commented on tools/rv_histogram.c:
```diff
@@ -179,8 +183,9 @@ void print_hist_stats(const rv_hist_t *stats, size_t stats_size)
 
         percent = ((float) freq / total_freq) * 100;
-        if (percent < 1.00)
+        if (percent < 1.00) {
```
Don't change this.

LambertWSJ commented on tools/rv_histogram.c:
```diff
@@ -241,39 +239,37 @@
+        bool has_insn = shdr->sh_flags & SHF_EXECINSTR;
 
-        if (!(shdr->sh_type & SHT_PROGBITS && shdr->sh_flags & SHF_EXECINSTR))
+        if (!(is_prg && has_insn))
```
@jserv 
For modifying a single line if statement, should I follow MISRA rules or [CONTRIBUTING.md](https://github.com/sysprog21/rv32emu/blob/master/CONTRIBUTING.md)?

jserv replied:
Follow [CONTRIBUTING.md](https://github.com/sysprog21/rv32emu/blob/master/CONTRIBUTING.md).

LambertWSJ commented on tools/rv_histogram.c:
```diff
@@ -145,34 +141,36 @@
 }
 
-static bool parse_args(int argc, char **args)
+static bool parse_args(int argc, char *args[])
```
@jserv 
The compiler will issue a warning if `args` is declared as a const array, Should I also change the `args` of the main function to `const char *args[]` in order to pass misra's rules?

jserv replied:
> The compiler will issue a warning if `args` is declared as a const array, Should I also change the `args` of the main function to `const char *args[]` in order to pass MISRA's rules?

I think so.

jserv commented on tools/rv_histogram.c:
```diff
@@ -241,39 +239,37 @@ int main(int argc, char *args[])
 #endif
             insn = *((uint32_t *) ptr);
-            ptr = (uintptr_t *) ((uint8_t *) ptr + 4);
+            ptr += 4;
```
Can you confirm if `ptr += 4` is equivalent to the original form?

LambertWSJ replied:
Theoretically, the results of the two are the same. By checking the number of loop executions and the histogram of the build/*.elf files, it is confirmed that they are the same as the original code.

BTW, `cc.elf` execution failed (core dump).

jserv replied:
> BTW, `cc.elf` execution failed (core dump).

Create a new issue on GitHub for tracking.

jserv commented on tools/rv_histogram.c:
```diff
@@ -241,39 +239,37 @@ int main(int argc, char *args[])
             if ((*((uint32_t *) ptr) & FC_OPCODE) != 0x3) {
                 insn = *((uint16_t *) ptr);
-                ptr = (uintptr_t *) ((uint8_t *) ptr + 2);
+                ptr += 2;
```
Can you confirm if `ptr += 2` is equivalent to the original form?

LambertWSJ replied:
ditto
